File: src/anneal.f
                
!#######################################################################
       SUBROUTINE anneal(criterio,p,svector,kmin,kmax,valores,vars,
     + bestval,bestvar,nfora,fora,ndentro,dentro,nsol,niter,
     + improvement,beta,temp,coolfreq,nqsi,qsi,esp,silog,solinit)
!#######################################################################

! Routine to perform a Simulated Annealing search for a k-variable subset
! of a set of p original variables, which maximizes one of several criteria.
! (The criteria currently considered are RM, RV and GCD). The user may force
!  the inclusion and/or exclusion of certain variables from the k-subset.
! 
! Designed to be called by an R/S function "anneal" (included in the package 
! "subselect", available from CRAN). 
!
! WARNING: Requires the LAPACK routines DSYEVD and DPOSV
! WARNING: Uses R's default Random Number Generator.
!
! INPUT : 
!
!criterio  - integer variable indicating which criterion of subset quality
!            was requested.
!   p      - integer variable indicating the number of original variables.
! svector  - double precision vector, giving the full covariance (or 
!            correlation) matrix of the p variables (given as a vector
!            for convenience in passing from R to Fortran). 
!kmin,kmax - integer variables, giving the smallest and largest cardinalities
!           of the subsets of variables that are wanted.
!  nfora   - integer variable indicating the number of original variables 
!            that are to be forcefully excluded from the k-subset.
!   fora   - integer vector, indicating the variable numbers associated
!            with the nfora variables that are to be excluded by force.
! ndentro  - integer variable indicating the number of original variables 
!            that are to be forcefully included in the k-subset.
!  dentro  - integer vector, indicating the variable numbers associated
!            with the ndentro variables that are to be included by force.
!  nsol    - integer indicating how many different solutions are to be 
!            produced by the simulated annealing search.
! niter    - integer variable, number of simulated annealing iterations 
!           requested for each solution.
!improvement- logical variable. If true, the best solution produced by 
!            Simulated Annealing will be subjected to a restricted local
!            improvement algorithm. (By default .true.).
!  beta    - double precision variable,(between 0 and 1) indicating the
!           geometric cooling factor for simulated annealing.
! temp    - double precision variable, indicating the initial temperature
!           to be used in the simulated annealing algorithm.
! coolfreq - integer indicating the cooling frequency, i.e., the number 
!           of iterations that are to be carried out for each fixed
!           temperature 
!  nqsi    - integer variable, indicating the dimension of the Principal 
!           Subspace with which the subspace spanned by the k-subset is
!            to be compared. (For GCD criterion only).
!  qsi     - integer vector, giving the ranks of the Principal Components 
!           which span the Principal Subspace described above (see nqsi).
!           (For GCD criterion only).
!  esp     - logical vector. If .true., indicates that the user has specified
!            a set of Principal Components different from the first k, to 
!            compare with the k-variable subset. (For GCD criterion only).
! silog    - logical variable indicating whether or not the user has specified
!             initial solutions.
! solinit  - integer variable, giving the user-specified initial solutions
!             (if any)

!
! OUTPUT: 
!
! valores  - double precision variable giving the final values produced by S.  
!            Annealing for each of the nsol solutions, in each cardinality, 
!            for the criterion required. Will pass to the R/S function "anneal"
!            for conversion to an nsol x (kmax-kmin+1) matrix within R/S, 
!            with each column associated with a  cardinality and each row with
!            a different solution. 
!  vars    - integer vector giving the list of variable numbers  belonging 
!            to  each of the nsol final subsets, for each cardinality (padded
!            with zeros if k<kmax). Will pass to the R/S function "anneal" for
!             conversion to an nsol x kmax x (kmax-kmin+1) 3-dimensional array.
! bestval - double precision vector with the best value of the criterion
!           obtained by running the routine, for each cardinality.  If 
!           improvement=.true. (the default), then the value given is the 
!           value resulting from running the restricted local search algorithm
!           on the best solution from Simulated Annealing (output for R/S).
! bestvar - integer vector with the variable numbers of the best subset 
!           obtained, for each cardinality.  If improvement=.true. (the 
!           default), then the subset given results from running the 
!           restricted local search algorithm on the best solution from 
!           Simulated Annealing (output for R/S).
!

! general declarations
       INTEGER p,criterio,niter,poriginal
       INTEGER fora(0:nfora),fica(0:p),dentro(0:ndentro),auxw(kmax)
       INTEGER vars(nsol*(kmax-kmin+1)*kmax)
       INTEGER bestvar((kmax-kmin+1)*kmax)
       LOGICAL setk(300),setkmax(300),improvement
       DOUBLE PRECISION s(300,300),sq(300,300),svector(p*p)
       DOUBLE PRECISION vmax,vactual,vcorrente
       DOUBLE PRECISION valores((kmax-kmin+1)*nsol)
       DOUBLE PRECISION beta,temp,critvalue
       integer coolfreq 
       DOUBLE PRECISION bestval(kmax-kmin+1)
       LOGICAL silog
       INTEGER solinit(kmax*nsol*(kmax-kmin+1))
! declarations only for the RM criterion
       DOUBLE PRECISION tracos,dobjrm
! declarations only for the RV criterion
       DOUBLE PRECISION tracosq,dobjrv
! declarations only for the GCD criterion
       DOUBLE PRECISION valp(p), vecp(300,300),dobjgcd
       INTEGER nqsi,qsi(p)
       LOGICAL esp


       external randsk1,dobjrm,dobjrv,dobjgcd
       external dcorrigesk,dannealing
       external dmelhoramentogen,inicializar


! initializations 
      call inicializar(criterio,p,s,svector,sq,nfora,fora,ndentro,
     +  dentro,fica,tracos,tracosq,valp,vecp,poriginal)


!*********************************
! The loop which is to be repeated for each cardinality of subsets, 
! between kmin e kmax starts here and ends together with the subroutine.

      do k=kmin,kmax

! Generates nsol solutions via Simulated Annealing

         if (criterio.eq.3) then
             if (.not.esp) then
               nqsi=k
               do m=1,nqsi
                 qsi(m)=m
               end do
             end if
         end if
         vmax=0

         do ksol=1,nsol

! determining the initial solution.

           if (silog) then
             do m=1,poriginal
               setk(m)=.false.
             end do           
             do m=1,k
               setk(solinit(nsol*kmax*(k-kmin)+nsol*(m-1)+ksol))=.true.
             end do
             if (nfora.ge.1) then
               do m=1,p
                 setk(m) = setk(fica(m))
               end do
             end if  
           else
               call randsk1(p-ndentro,k-ndentro,setk)           
               if(ndentro>0) call dcorrigesk(ndentro,dentro,p,setk)
           endif



           if (criterio.eq.1) then
               vactual=dobjrm(k,setk,p,s,sq)
           end if
           if (criterio.eq.2) then
               vactual=dobjrv(k,setk,p,s,sq)
           end if
           if (criterio.eq.3) then
               vactual=dobjgcd(nqsi,qsi,valp,vecp,k,setk,p,s,fica)
           end if
        
           call dannealing(criterio,p,k,s,sq,setk,vactual,ndentro,
     +       dentro,niter,beta,temp,coolfreq,tracos,tracosq,
     +       nqsi,qsi,valp,vecp,fica)

           if (criterio.eq.1) then
              critvalue = dsqrt(vactual/tracos)
           end if
           if (criterio.eq.2) then
              critvalue = dsqrt(vactual/tracosq)
           end if
           if (criterio.eq.3) then
              critvalue = vactual/dsqrt(dfloat(nqsi*k))
           end if

! preparing the output for R/S
            valores((k-kmin)*nsol+ksol)=critvalue
            jjaux=0
           do i=1,p
             if (setk(i)) then
               jjaux=jjaux+1
               vars(nsol*kmax*(k-kmin)+nsol*(jjaux-1)+ksol) = fica(i)
             end if
           end do  

           if(vactual>vmax) then
              vmax=vactual
              do j=1,p
                 setkmax(j)=setk(j)
              end do
           end if
         end do
         naux=0
         do j=1,p
           if(setkmax(j)) then
              naux=naux+1
              auxw(naux)= fica(j)
           end if
         end do
         if (criterio.eq.1) then
            critvalue=dsqrt(vmax/tracos)
         end if
         if (criterio.eq.2) then
            critvalue=dsqrt(vmax/tracosq)
         end if
         if (criterio.eq.3) then
              critvalue = vmax/dsqrt(dfloat(nqsi*k))
         end if


! If improvement is "true" runs a restricted local improvement algorithm
! on the best of the nsol solutions produced by Simulated Annealing.  

         if (improvement) then
            vcorrente = vmax
            call dmelhoramentogen(criterio,p,setkmax,vmax,ndentro,
     +        dentro,k,s,sq,nqsi,qsi,valp,vecp,fica)
            if (vmax > vcorrente) then 
                   naux=0
                   do j=1,p
                     if(setkmax(j)) then
                       naux=naux+1
                       auxw(naux)= fica(j)
                     end if
                   end do
                   if (criterio.eq.1) then
                       critvalue=dsqrt(vmax/tracos)
                   end if
                   if (criterio.eq.2) then
                       critvalue=dsqrt(vmax/tracosq)
                   end if
                   if (criterio.eq.3) then
                       critvalue = vmax/dsqrt(dfloat(nqsi*k))
                   end if

            end if      
         end if

! End of restricted improvement

         bestval(k-kmin+1)=critvalue
         do j=1,k
            bestvar(kmax*(k-kmin)+j)=auxw(j)
         end do
       end do  

! End of the k-loop

      return
      end

File: R/anneal.R
anneal<-function(mat, kmin, kmax=kmin, nsol=1, niter=1000,
exclude=NULL, include=NULL, improvement=TRUE, 
setseed = FALSE, cooling=0.05, temp=1,
coolfreq=1, criterion="RM", pcindices=1:kmax, initialsol=c(0)){


# validation of input

        if (!is.matrix(mat)) 
          stop("Data is missing or is not given in matrix form")
        if (dim(mat)[1] != dim(mat)[2]) {
          mat <- var(mat)
          warning("Data must be given as a covariance or correlation matrix.\n It has been assumed that you wanted the covariance matrix of the \n data matrix supplied")}
        if (kmax < kmin) {
          aux<-kmin
          kmin<-kmax
          kmax<-aux
          warning("Argument kmin should precede argument kmax.\n Since the value of kmin exceeded that of kmax, they have been swapped \n")}
        if (!(as.integer(coolfreq) == coolfreq) | (coolfreq < 1)) stop("\n The cooling frequency must be a non-negative integer")
        p<-dim(mat)[[2]]
        if (kmin >= p) {
             kmin<-p-1
             warning("\n The value of kmin requested is equal to or exceeds the number \n of variables. It has been set at p-1")
            }  
        if (kmax >= p) {
             kmax<-p-1
             warning("\n The value of kmax requested is equal to or exceeds the number \n of variables. It has been set at p-1")
            }
        if (sum(duplicated(c(exclude,include))) > 0)
          stop("\n You have requested that the same variable be both included in, and excluded \n from, the subset.")
         labelsrm<-c("RM","Rm","rm","1",1)
         labelsrv<-c("RV","Rv","rv","2",2)
         labelsgcd<-c("GCD","Gcd","gcd","3",3)
         if (sum(criterion == c(labelsrm,labelsrv,labelsgcd)) == 0) stop("criterion requested is not catered for, or has been misspecified")
         

# initializations for the Fortran subroutine
       
        if (setseed == TRUE) {set.seed(2,kind="default")} 
        if (sum(criterion == labelsrm) > 0) criterio<-1
        if (sum(criterion == labelsrv) > 0) criterio<-2
        if (sum(criterion == labelsgcd) > 0) criterio<-3

        nexclude<-length(exclude)
        if (kmax >= p-nexclude) stop("\n Cardinalities requested are too large for the requested number of excluded variables")
        if (nexclude !=0) exclude<-sort(exclude)  
        exc<-c(0,exclude)
        ninclude<-length(include)
        if (kmin <= ninclude) stop("\n Cardinalities requested are too small for the requested number of included variables")
        if (ninclude !=0) include<-sort(include)
        inc<-c(0,include)
        if (cooling<=0 || cooling >=1) stop("\n values of cooling must be between 0 and 1")
        if (length(pcindices) != kmax) esp<-TRUE else {if (pcindices
        != 1:kmax) esp<-TRUE  else esp<-FALSE}

# initializations when initial solutions are specified; checking for
# nature of initialsol (and how to interpret it) and for
# conflicts with the exclude and include parameters (which inital
# solutions must respect)

        if (initialsol == c(0)) {silog <- FALSE}
        else {silog <- TRUE  # initial solutions have been specified by user          

# checking for the presence of variables that are to be forcefully excluded
            if ((nexclude != 0) & (sum(exclude == rep(initialsol,rep(length(exclude),length(as.vector(initialsol))))) !=0)) stop("\n the specified initial solutions contain variables that are to be excluded")

# how to deal with various formats of input (of initial solutions) for
# a single cardinality

          dimsol<-dim(initialsol)
          if (length(dimsol) > 3) {
            stop("\n Can't handle arrays of more than 3 dimensions")}
          if (kmin == kmax) {

# inital solution is a vector: must be repeated if nsol > 1

           if (is.vector(initialsol)) {
            if (length(initialsol) != kmax) 
             {stop("\n The specified initial and final solutions have different cardinalities")}
            else
             {if (nsol > 1) initialsol<-matrix(nrow=nsol,ncol=kmax,rep(initialsol,rep(nsol,kmax)))}
           }

# initial solution is array?

           if (is.array(initialsol)) {

# initialsol is 3-d array
           if (length(dimsol) == 3) {
            if (dim(initialsol)[[3]] > 1) stop("\n The input array of initial solutions must have dimensions as \n (1 or nsol) x k x 1 when a single cardinality is requested")
            else initialsol<-matrix(nrow=dim(initialsol)[[1]],ncol=dim(initialsol)[[2]],initialsol)
            }
           else

# initialsol is matrix: must be of form nsol x k
 
            {if (dim(initialsol)[[2]] != kmax) stop("\n Input matrix of initial solutions must have as many columns as variables in the requested subset")
             else 
              {if  (dim(initialsol)[[1]] != nsol) {
                if (dim(initialsol)[[1]] == 1) {initialsol<-matrix(nrow=nsol,ncol=dim(initialsol)[[2]],rep(initialsol,rep(nsol,kmax))) }
                else {stop("\n The number of initial solutions can only be 1 or nsol (number of final solutions requested)")}}}}
              }}

# how to deal with various formats of input (initialsol) if more than
# one cardinality is requested

             else  # (if kmax > kmin), i.e., more than one cardinality requested

# initial solution cannot be a single vector

              {if (is.vector(initialsol)) stop("\n There must be initial solutions for all cardinalities requested")

               if (is.array(initialsol)) {

# initial solution 3-d array?

                  if (length(dim(initialsol)) == 3) {
                   if ((dim(initialsol)[[3]] != length(kmin:kmax)) | (dim(initialsol)[[2]] != kmax) | ((dim(initialsol)[[1]] != nsol) & (dim(initialsol)[[1]] > 1))) stop("\n The input array of initial solutions must have dimensions as \n (1 or nsol) x kmax x no. of different cardinalities requested")
                   else 
                    { if (dim(initialsol)[[1]] != nsol) initialsol<-array(dim=c(nsol,kmax,length(kmin:kmax)),rep(initialsol,rep(nsol,kmax*length(kmin:kmax))))
                  }}
                else 

# if initalsol is a matrix

                    {
                    if ((dim(initialsol)[[2]] != kmax) | (dim(initialsol)[[1]] != length(kmin:kmax))) stop("\n A matrix of initial solutions for more than one cardinality must be of type no. of cardinalities x kmax")
                    if (nsol > 1) initialsol<-array(dim=c(nsol,kmax,length(kmin:kmax)),rep(t(initialsol),rep(nsol,kmax*length(kmin:kmax))))
}}}
                 if ((ninclude != 0) & (sum(include == rep(initialsol,rep(length(include),length(as.vector(initialsol))))) != nsol*length(kmin:kmax)*length(include))) stop("\n Not all the specified initial solutions contain the variables that are to be included") 
}

#
# initializations for Fortran output
#

        valores<-rep(0.0,length(kmin:kmax)*nsol)    
        vars<-rep(0,nsol*length(kmin:kmax)*kmax)
        bestval<-rep(0.0,length(kmin:kmax))
        bestvar<-rep(0,kmax*length(kmin:kmax))

#
# call to the Fortran subroutine
#

        Fortout<-.Fortran("anneal",as.integer(criterio),as.integer(p),
          as.double(as.vector(mat)),
          as.integer(kmin),as.integer(kmax),as.double(valores),
          as.integer(vars),as.double(bestval),as.integer(bestvar),
          as.integer(nexclude),as.integer(exc),as.integer(ninclude),
          as.integer(inc),as.integer(nsol),as.integer(niter),
          as.logical(improvement),
          as.double(cooling),as.double(temp),
          as.integer(coolfreq),as.integer(length(pcindices)),
          as.integer(pcindices),as.logical(esp),as.logical(silog),
          as.integer(as.vector(initialsol)),PACKAGE="subselect")

# preparing the output

        valores<-matrix(nrow=nsol,ncol=length(kmin:kmax),Fortout[[6]])
        dimnames(valores)<-list(paste("Solution",1:nsol,sep=" "),paste("card.",kmin:kmax,sep=""))
        variaveis<-array(Fortout[[7]],c(nsol,kmax,length(kmin:kmax)))
        dimnames(variaveis)<-list(paste("Solution",1:nsol,sep=" "),paste("Var",1:kmax,sep="."),paste("Card",kmin:kmax,sep="."))
        bestval<-Fortout[[8]]
        names(bestval)<-paste("Card",kmin:kmax,sep=".")
        bestvar<-t(matrix(nrow=kmax,ncol=length(kmin:kmax),Fortout[[9]]))
        dimnames(bestvar)<-list(paste("Card",kmin:kmax,sep="."),paste("Var",1:kmax,sep="."))
        output<-list(variaveis,valores,bestval,bestvar)
        names(output)<-c("subsets","values","bestvalues","bestsets")
        output}


